webapi/setaltsig.go

```diff
@@ -0,0 +1,142 @@
+
+	const funcName = "setAltSig"
+
+	// Get values which have been added to context by middleware.
+	dcrdClient := c.MustGet("DcrdClient").(Node)
+	reqBytes := c.MustGet("RequestBytes").([]byte)
+	commitmentAddr := c.MustGet("CommitmentAddress").(string)
+
+	if cfg.VspClosed {
+		sendError(errVspClosed, c)
+		return
+	}
+
+	var request SetAltSigRequest
+	if err := binding.JSON.BindBody(reqBytes, &request); err != nil {
+		log.Tracef("%s: Bad request (clientIP=%s): %v", funcName, c.ClientIP(), err)
+		sendErrorWithMsg(err.Error(), errBadRequest, c)
+		return
+	}
+
+	altSigAddr, ticketHash := request.AltSigAddress, request.TicketHash
+
+	currentData, err := db.AltSigData(ticketHash)
+	if err != nil {
+		log.Errorf("%s: db.AltSigData (ticketHash=%s): %v", funcName, ticketHash, err)
+		sendError(errInternalError, c)
+		return
+	}
+	if currentData != nil {
+		msg := "alternate signature data already exists"
+		log.Errorf("%s: %s (ticketHash=%s)", funcName, msg, ticketHash)
+		sendErrorWithMsg(msg, errBadRequest, c)
+		return
+
+	}
+
+	// Fail fast if the pubkey doesn't decode properly.
+	addr, err := stdaddr.DecodeAddressV0(altSigAddr, cfg.NetParams)
+	if err != nil {
+		log.Tracef("%s: Alt sig address cannot be decoded (clientIP=%s): %v", funcName, c.ClientIP(), err)
+		sendErrorWithMsg(err.Error(), errBadRequest, c)
+		return
+	}
+	if _, ok := addr.(*stdaddr.AddressPubKeyHashEcdsaSecp256k1V0); !ok {
+		log.Tracef("%s: Alt sig address is unexpected type (clientIP=%s type=%T): %v", funcName, c.ClientIP(), addr, err)
+		sendErrorWithMsg("wrong type for alternate signing address", errBadRequest, c)
+		return
+	}
+
+	// Get ticket details.
+	rawTicket, err := dcrdClient.GetRawTransaction(ticketHash)
+	if err != nil {
+		log.Errorf("%s: dcrd.GetRawTransaction for ticket failed (ticketHash=%s): %v", funcName, ticketHash, err)
+		sendError(errInternalError, c)
+		return
+	}
+
+	// Ensure this ticket is eligible to vote at some point in the future.
+	canVote, err := dcrdClient.CanTicketVote(rawTicket, ticketHash, cfg.NetParams)
+	if err != nil {
+		log.Errorf("%s: dcrd.CanTicketVote error (ticketHash=%s): %v", funcName, ticketHash, err)
+		sendError(errInternalError, c)
+		return
+	}
+	if !canVote {
+		log.Tracef("%s: unvotable ticket (clientIP=%s, ticketHash=%s)",
+			funcName, c.ClientIP(), ticketHash)
+		sendError(errTicketCannotVote, c)
+		return
+	}
+
+	sigStr := c.GetHeader("VSP-Client-Signature")
+
+	res := SetAltSigResponse{
+		Timestamp: time.Now().Unix(),
+		Request:   reqBytes,
+	}
+	resBytes, err := json.Marshal(res)
+	if err != nil {
+		log.Errorf("JSON marshal error: %v", err)
+		sendError(errInternalError, c)
+		return
+	}
+
+	sig := ed25519.Sign(signPrivKey, resBytes)
+	serverSigStr := base64.StdEncoding.EncodeToString(sig)
+	c.Writer.Header().Set("VSP-Server-Signature", serverSigStr)
+
+	c.AbortWithStatusJSON(http.StatusOK, res)
+
+	data := &database.AltSigData{
+		AltSigAddr: altSigAddr,
+		ReqSigAddr: commitmentAddr,
```

Comment:
The commitment address was used to sign Req which adds the alternate signature address. It must be known in order to prove that the user added an alternate signature.

> all we really need for vspd to function is a mapping between ticketHash and altSig

This is true as long as we can fish out the commitment address later. Since I believe we always can, we can remove this. It should be in the ticket hex in the request.

Reply:
Removed in 0441516 and ce1b625